### PR TITLE
New performance test: ICS space amplification

### PR DESCRIPTION
**Pull request by yarongilor**
This PR is part of: https://trello.com/c/SlMpIboP/635-incremental-compaction?menu=filter&filter=member:yarongilor2

pipeline job is running here: https://jenkins.scylladb.com/job/scylla-staging/job/longevity-tests-ICS/job/ics_space_amplification_test/

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by fruch** [comment]
@yarongilor can you rephrase the title ? i.e. is this a new test ? a new nemesis ?

File: ics_space_amplification_test.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+        :rtype: dictionary with capacity per node-ip
+        """
+        dict_nodes_used_capacity = {}
+        for node in self.db_cluster.nodes:
+            dict_nodes_used_capacity[node.private_ip_address] = self._get_used_capacity_gb(node=node)
+        return dict_nodes_used_capacity
+
+    @retrying(n=80, sleep_time=60, allowed_exceptions=(AssertionError,))
+    def wait_no_compactions_running(self):
```
Comment (fruch) [inline ics_space_amplification_test.py]: how this is different from `wait_no_compactions_running()` in `tester.py`

File: ics_space_amplification_test.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        num_of_columns = 5
+        ops_num = 200200300
+        overwrite_ops_num = ops_num // 2
+        total_new_data_to_write_gb = ops_num * column_size * num_of_columns / (1024 ** 3)
+        total_data_to_overwrite_gb = overwrite_ops_num * column_size * num_of_columns / (1024 ** 3)
+        keyspace_num = 1
+
+        self.log.debug('Test Space-amplification on writing new data')
+        prepare_write_cmd = "cassandra-stress write cl=ALL n={ops_num}  -schema 'replication(factor=3)" \
+                            " compaction(strategy=IncrementalCompactionStrategy)' -port jmx=6868 -mode cql3 native" \
```
Comment (fruch) [inline ics_space_amplification_test.py]: I thought that cassandra-stress doesn't support this value....

File: ics_space_amplification_test.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    def test_ics_space_amplification(self):  # pylint: disable=too-many-locals
+        # too-many-branches,too-many-statements
+        self._pre_create_schema_with_compaction()
+        stress_queue = list()
+        write_queue = list()
+        verify_queue = list()
+        column_size = 200
+        num_of_columns = 5
+        ops_num = 200200300
```
Comment (fruch) [inline ics_space_amplification_test.py]: some comment on what is this number, and how it was calculated or agreed. would help to understand what's going on

File: ics_space_amplification_test.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        overwrite_ops_num = ops_num // 2
+        total_new_data_to_write_gb = ops_num * column_size * num_of_columns / (1024 ** 3)
+        total_data_to_overwrite_gb = overwrite_ops_num * column_size * num_of_columns / (1024 ** 3)
+        keyspace_num = 1
+
+        self.log.debug('Test Space-amplification on writing new data')
+        prepare_write_cmd = "cassandra-stress write cl=ALL n={ops_num}  -schema 'replication(factor=3)" \
+                            " compaction(strategy=IncrementalCompactionStrategy)' -port jmx=6868 -mode cql3 native" \
+                            " -rate threads=1000 -col 'size=FIXED({column_size}) n=FIXED(num_of_columns)'" \
+                            " -pop seq=1..200200300 -log interval=15".format(**locals())
```
Comment (fruch) [inline ics_space_amplification_test.py]: if there a variable `ops_num` maybe it can be used here ?

File: ics_space_amplification_test.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        ops_num = 200200300
+        overwrite_ops_num = ops_num // 2
+        total_new_data_to_write_gb = ops_num * column_size * num_of_columns / (1024 ** 3)
+        total_data_to_overwrite_gb = overwrite_ops_num * column_size * num_of_columns / (1024 ** 3)
+        keyspace_num = 1
+
+        self.log.debug('Test Space-amplification on writing new data')
+        prepare_write_cmd = "cassandra-stress write cl=ALL n={ops_num}  -schema 'replication(factor=3)" \
+                            " compaction(strategy=IncrementalCompactionStrategy)' -port jmx=6868 -mode cql3 native" \
+                            " -rate threads=1000 -col 'size=FIXED({column_size}) n=FIXED(num_of_columns)'" \
```
Comment (fruch) [inline ics_space_amplification_test.py]: make sure you select loader which are strong enough for this

File: test-cases/longevity/ics_space_amplification_test.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+test_duration: 1500
+
+round_robin: true
+
+n_db_nodes: 4
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
+instance_type_loader: 'c4.2xlarge'
```
Comment (fruch) [inline test-cases/longevity/ics_space_amplification_test.yaml]: A note, on why using bigger loaders might help, so people won't think it's by mistake or something like that (and to justify the pricing in a way)

File: jenkins-pipelines/ics_space_amplification_test.jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!groovy
```
Comment (fruch) [inline jenkins-pipelines/ics_space_amplification_test.jenkinsfile]: since this is not a longevity test, maybe would be useful to prefix this filename with features- or at least ICS (i.e with capital letters)

File: ics_space_amplification_test.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        now = time.time()
+        results = self.prometheus_db.query(query=compaction_query, start=now - 60, end=now)
+        self.log.debug(f"scylla_compaction_manager_compactions: {results}")
+        # if all are zeros the result will be False, otherwise there are still compactions
+        if results:
+            assert any([float(v[1]) for v in results[0]["values"]]) is False, \
+                "Waiting until all compactions settle down"
+
+    def test_ics_space_amplification(self):  # pylint: disable=too-many-locals
+        # too-many-branches,too-many-statements
```
Comment (fruch) [inline ics_space_amplification_test.py]: please add a docstring to explian what this test is trying todo, what is the goals of it
(maybe it's copy-paste from the trello task, but it should be in the code)

File: ics_space_amplification_test.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        verify_nodes_space_amplification(dict_nodes_space_amplification=dict_nodes_space_amplification)
+
+
+def verify_nodes_space_amplification(dict_nodes_space_amplification):
+    for node_ip, space_amplification_gb in dict_nodes_space_amplification.items():
+        assert space_amplification_gb < MAX_ICS_SPACE_AMPLIFICATION_ALLOWED_GB, \
+            f'Node {node_ip} space amplification of: {space_amplification_gb} exceeds the maximum allowed ({MAX_ICS_SPACE_AMPLIFICATION_ALLOWED_GB})'
+
+
+def generate_capacity_query_postfix(node):
```
Comment (fruch) [inline ics_space_amplification_test.py]: this sound more like `limit_query_to_specific_node()`

File: ics_space_amplification_test.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            "The used filesystem capacity on node {} is: {} MB/ {} GB".format(node.public_ip_address, used_size_mb,
+                                                                              used_size_gb))
+        return used_size_gb
+
+    def _get_filesystem_available_size_list(self, node, start_time):
+        """
+        :param node:
+        :return:
+        """
+        available_capacity_query = f'{AVAIL_SIZE_METRIC}{generate_capacity_query_postfix(node)}'
```
Comment (fruch) [inline ics_space_amplification_test.py]: Please don't use f-string if this work is intend to backported to 2019.1

File: ics_space_amplification_test.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        available_size_res = self.prometheus_db.query(query=available_capacity_query, start=int(start_time),
+                                                      end=int(time.time()))
+        assert available_size_res, "No results from Prometheus"
+        if not available_size_res[0]:  # if no returned values - try the old metric names.
+            available_capacity_query = f'{AVAIL_SIZE_METRIC_OLD}{generate_capacity_query_postfix(node)}'
```
Comment (fruch) [inline ics_space_amplification_test.py]: Please don't use f-string if this work is intend to backported to 2019.1

File: ics_space_amplification_test.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        return available_size_res[0]["values"]
+
+    def _get_filesystem_total_size_gb(self, node):
+        """
+        :param node:
+        :return:
+        """
+        fs_size_metric = 'node_filesystem_size_bytes'
+        fs_size_metric_old = 'node_filesystem_size'
+        filesystem_capacity_query = f'{fs_size_metric}{generate_capacity_query_postfix(node)}'
```
Comment (fruch) [inline ics_space_amplification_test.py]: Please don't use f-string if this work is intend to backported to 2019.1

File: ics_space_amplification_test.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        fs_size_res = self.prometheus_db.query(query=filesystem_capacity_query, start=int(time.time()) - 5,
+                                               end=int(time.time()))
+        assert fs_size_res, "No results from Prometheus"
+        if not fs_size_res[0]:  # if no returned values - try the old metric names.
+            filesystem_capacity_query = f'{fs_size_metric_old}{generate_capacity_query_postfix(node)}'
```
Comment (fruch) [inline ics_space_amplification_test.py]: Please don't use f-string if this work is intend to backported to 2019.1

File: ics_space_amplification_test.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+        assert used_cap_res, "No results from Prometheus"
+        used_size_mb = float(used_cap_res[0]["values"][0][1]) / float(MB_SIZE)
+        used_size_gb = float(used_size_mb / 1024)
+        self.log.debug(
+            "The used filesystem capacity on node {} is: {} MB/ {} GB".format(node.public_ip_address, used_size_mb,
+                                                                              used_size_gb))
+        return used_size_gb
+
+    def _get_filesystem_available_size_list(self, node, start_time):
+        """
```
Comment (fruch) [inline ics_space_amplification_test.py]: if you are adding a docstring to a function, please write something in it.
and make sure it covers all the arguments.

if you choose not to document it, remove docstring.

**Comment by fruch** [comment]
@yarongilor, a general note, you are doing lots of calculations base on queries here, comments in the beginning of each function to describe it, would help to give it context. 

cause reviewing the code only, you can'y understand what the intent, and what this calculation is trying to achieve, so it's kind of impossible to asses this code. for example if this is something relevant to other test or even if we already have this type of calculation in our code.

File: test-cases/features/ics_space_amplification_test.yaml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+nemesis_class_name: 'NoOpMonkey'
+user_prefix: 'ics_space_amplification'
+server_encrypt: false
+client_encrypt: false
+
+
+use_mgmt: false
+nemesis_during_prepare: false
+
+append_scylla_args: '--compaction-enforce-min-threshold 1'
```
Comment (fruch) [inline test-cases/features/ics_space_amplification_test.yaml]: since this overwrite the defaults we have in `test_defaults.yaml`, double check if you don't want to include that too

Comment (yarongilor) [reply test-cases/features/ics_space_amplification_test.yaml]: that's may be right. i can't tell yet since i need a clarification about this argument from @raphaelsc

File: ics_space_amplification_test.py (hunk 1 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: it is different in: (1) n=80  (2) "if results: assert..."

File: ics_space_amplification_test.py (hunk 2 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: it should be fixed by a recent PR of @bhalevy

File: ics_space_amplification_test.py (hunk 3 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: done

File: jenkins-pipelines/ics_space_amplification_test.jenkinsfile (hunk 7 above)
Comment (yarongilor) [reply jenkins-pipelines/ics_space_amplification_test.jenkinsfile]: done

File: ics_space_amplification_test.py (hunk 8 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: done

File: ics_space_amplification_test.py (hunk 9 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: renamed to 'generate_node_capacity_query_postfix'

File: ics_space_amplification_test.py (hunk 10 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: done

File: ics_space_amplification_test.py (hunk 14 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: done

**Comment by yarongilor** [comment]
> help to giv

added some description comments to variables and methods.

File: ics_space_amplification_test.py (hunk 1 above)
Comment (fruch) [reply ics_space_amplification_test.py]: that's not enough diffrence to justify two function doing the same thing, and with the same name.
refactor to function if needed to add needed functionality from it. or at least warp it with with needed code.

File: ics_space_amplification_test.py (hunk 2 above)
Comment (fruch) [reply ics_space_amplification_test.py]: so this PR should also also should have the changes need to use this new c-s version, i.e. it's need a new AMIs with it. @amoskong might be able to help once it's ready.

Comment (yarongilor) [reply ics_space_amplification_test.py]: perhaps we better split it to 2 PRs. this is also since PR: https://github.com/scylladb/scylla-enterprise-tools-java/pull/18   is not backported to 2019.1 yet.

File: test-cases/features/ics_space_amplification_test.yaml (hunk 15 above)
Comment (fruch) [reply test-cases/features/ics_space_amplification_test.yaml]: I've meant the other argument we put as default. You might want them too, right now only this argument would be used.

**Comment by fruch** [comment]
@yarongilor 
check the precommit output
we fixed this typo in the master (and all branches that it was)
```
14:48:36  ************* Module ics_space_amplification_test
14:48:36  ics_space_amplification_test.py:4:0: E0611: No name 'IcsLongevetyTest' in module 'longevity_ics_test' (no-name-in-module)
14:48:36  
14:48:36  -------------------------------------------------------------------
```

Also please tag the backports for this PR

File: test-cases/features/ics_space_amplification_test.yaml (hunk 15 above)
Comment (raphaelsc) [reply test-cases/features/ics_space_amplification_test.yaml]: yaron, I think it's not good to have that flag turned on by default. Users will likely not use it. So better test ICS with defaults, no? We have plans to improve that, but by now, I think we should leave that flag turned off as users will likely do so too.

File: ics_space_amplification_test.py (hunk 1 above)
Comment (yarongilor) [reply ics_space_amplification_test.py]: fixed in tester.py

File: test-cases/features/ics_space_amplification_test.yaml (hunk 15 above)
Comment (yarongilor) [reply test-cases/features/ics_space_amplification_test.yaml]: @raphaelsc , that make sense. the consequence will be the test fails on first step. so do we consider that a real issue that should be handled in a github issue?

Comment (yarongilor) [reply test-cases/features/ics_space_amplification_test.yaml]: in any case, i opened task: https://trello.com/c/6LpWsBON/1437-turn-off-appendscyllaargs-compaction-enforce-min-threshold-1 .
so this PR could be meanwhile concluded and merged.

Comment (fruch) [reply test-cases/features/ics_space_amplification_test.yaml]: @yarongilor you are still missing my comment here, that the other commandline argument we add as default in out tests are overwritten by this line.
please make that is the intended effect you want,

look at the defaults, when you set this in your yaml, only what configured in you yaml for this value is used
https://github.com/scylladb/scylla-cluster-tests/blob/9ddb0e6f6b4fabf4ae7eaf1c3029cf51fcc1a0da/defaults/test_default.yaml#L13

Comment (yarongilor) [reply test-cases/features/ics_space_amplification_test.yaml]: > add as default in out tests are overwritten by this line.
> please make that is the intended effect you want,
> 
> look at the defaults, when you set this in your yaml, only what configured in you yaml for this value is used

i guess there is something misleading here with the naming, since this is, of course, not intended to override other default, but only to add one. so the name "append_scylla_args" causes a confusion. it doesn't append but rather overrides the default. i would suggest to change it so that it will actually append and given value to the default values list as long as it's not contained in default list already. perhaps for an override option we can add a new parameter named "set_scylla_args".

Comment (yarongilor) [reply test-cases/features/ics_space_amplification_test.yaml]: for this PR, the flag is now removed: 
06c3c14f - (HEAD -> ics_space_amplification, origin/ics_space_amplification) Removed append_scylla_args: '--compaction-enforce-min-threshold 1

**Comment by yarongilor** [comment]
> ur code.



> @yarongilor
> check the precommit output
> we fixed this typo in the master (and all branches that it was)
> 
> ```
> 14:48:36  ************* Module ics_space_amplification_test
> 14:48:36  ics_space_amplification_test.py:4:0: E0611: No name 'IcsLongevetyTest' in module 'longevity_ics_test' (no-name-in-module)
> 14:48:36  
> 14:48:36  -------------------------------------------------------------------
> ```
> 
> Also please tag the backports for this PR

already fixed in previous commit.

**Comment by fruch** [comment]
now you have a new pre-commit issue, code duplication:

```
12:05:22  ************* Module unit_tests.test_version_utils
12:05:22  unit_tests/test_version_utils.py:1:0: R0801: Similar lines in 2 files
12:05:22  ==ics_space_amplification_test:27
12:05:22  ==longevity_ics_test:14
12:05:22          self.log.debug('Pre Creating Schema for c-s with {} keyspaces'.format(keyspace_num))
12:05:22          for i in range(1, keyspace_num + 1):
12:05:22              keyspace_name = 'keyspace{}'.format(i)
12:05:22              self.create_keyspace(keyspace_name=keyspace_name, replication_factor=3)
12:05:22              self.log.debug('{} Created'.format(keyspace_name))
12:05:22              self.create_table(name='standard1', keyspace_name=keyspace_name, key_type='blob', read_repair=0.0,
12:05:22                                compact_storage=True,
12:05:22                                columns=DICT_DEFAULT_BLOB_COLUMNS, compaction=compaction, compression=compression,
12:05:22                                scylla_encryption_options=scylla_encryption_options)
12:05:22          if create_all_keyspaces_table: (duplicate-code)
12:05:22  
12:05:22  -------------------------------------------------------------------
12:05:22  Your code has been rated at 10.00/10 (previous run: 9.99/10, +0.01)
12:05:22  
```

**Comment by yarongilor** [comment]
> now you have a new pre-commit issue, code duplication:
> 
> ```
> 12:05:22  ************* Module unit_tests.test_version_utils
> 12:05:22  unit_tests/test_version_utils.py:1:0: R0801: Similar lines in 2 files
> 12:05:22  ==ics_space_amplification_test:27
> 12:05:22  ==longevity_ics_test:14
> 12:05:22          self.log.debug('Pre Creating Schema for c-s with {} keyspaces'.format(keyspace_num))
> 12:05:22          for i in range(1, keyspace_num + 1):
> 12:05:22              keyspace_name = 'keyspace{}'.format(i)
> 12:05:22              self.create_keyspace(keyspace_name=keyspace_name, replication_factor=3)
> 12:05:22              self.log.debug('{} Created'.format(keyspace_name))
> 12:05:22              self.create_table(name='standard1', keyspace_name=keyspace_name, key_type='blob', read_repair=0.0,
> 12:05:22                                compact_storage=True,
> 12:05:22                                columns=DICT_DEFAULT_BLOB_COLUMNS, compaction=compaction, compression=compression,
> 12:05:22                                scylla_encryption_options=scylla_encryption_options)
> 12:05:22          if create_all_keyspaces_table: (duplicate-code)
> 12:05:22  
> 12:05:22  -------------------------------------------------------------------
> 12:05:22  Your code has been rated at 10.00/10 (previous run: 9.99/10, +0.01)
> 12:05:22  
> ```

thanks @fruch , fixed.
